keps/sig-storage/2644-honor-pv-reclaim-policy/README.md

```diff
@@ -0,0 +1,421 @@
+			// Check if there is a request to delete the volume explicitly.
+			if !ctrl.checkFinalizer(volume, pvDeletionProtectionFinalizer) {
+				return false
+			} else {
+				klog.Info("The volume %s was explicitly requested to be deleted from underlying storage", volume.Name)
+			}
+		}
+	}
+// Remaining code below skipped for brevity...
+// Please refer sig-storage-lib-external-provisioner/controller/controller.go for the full code.
+```
+When the `shouldDelete` checks succeed, a delete volume request is initiated on the driver. This ensures that the volume
+is deleted in backend. 
+
+Once the volume is deleted from the backend, the finalizer can be removed. This allows the pv to be removed from the api
+server.
+
+```go
+// deleteVolumeOperation attempts to delete the volume backing the given
+// volume. Returns error, which indicates whether deletion should be retried
+// (requeue the volume) or not
+func (ctrl *ProvisionController) deleteVolumeOperation(ctx context.Context, volume *v1.PersistentVolume) error {
+	operation := fmt.Sprintf("delete %q", volume.Name)
+	klog.Info(logOperation(operation, "started"))
+	//....
+	//....
+	//....
+    klog.Infof("deleteVolumeOperation succeeded for the PV: %s", volume.Name)
+    klog.Info(logOperation(operation, "volume deleted"))
+    klog.Infof("deleteVolumeOperation attempting to remove the pv deletion protection finalizer for the PV: %s", volume.Name)
+    if ctrl.checkFinalizer(volume, pvDeletionProtectionFinalizer) {
+        // If there's a finalizer remove it to allow the pv deletion from the api server.
+        if err = ctrl.removePvDeletionProtectionFinalizer(volume); err != nil {
+            klog.Info(logOperation(operation, "failed to remove persistentvolume deletion protection finalizer: %v", err))
+            return err
+        }
+    }
+    //...
+    //...
+// Remaining code below skipped for brevity...
+// Please refer sig-storage-lib-external-provisioner/controller/controller.go for the full code.
+```
+
+### Test Plan
+
+#### E2E tests
+An e2e test to exercise deletion of PV prior to deletion of PVC for a `Bound` PV-PVC pair.
+
+
+### Graduation Criteria
+
+#### Alpha
+* Feedback
+* Fix of the issue behind a feature flag.
+* Unit tests and e2e tests outlined in design proposal implemented.
+
+#### Alpha -> Beta
+* Allow the Alpha fix to soak for one release.
+
+#### Beta -> GA
+* Deployed in production and have gone through at least one K8s upgrade.
+
+### Upgrade / Downgrade Strategy
+* Upgrade from old Kubernetes(1.22) to new Kubernetes(1.23) with `HonorPvReclaim` flag enabled
+
+In the above scenario, the upgrade will cause change in default behavior as described in the current KEP. Additionally,
+if there are PVs that have a valid associated PVC and deletion timestamp set, then a finalizer is added to the PV.
+
+* Downgrade from new Kubernetes(1.22) to old Kubernetes(1.21).
+
+In the above scenario, downgrade will retain the current behavior.
```

Comment:
There may be PVs with the deletion finalizer that the older Kubernetes does not remove. Such PVs will be in the API server forever.